cachi2/core/package_managers/rpm/main.py

```diff
@@ -0,0 +1,204 @@
+    """Process the input lockfile, fetch packages and generate SBOM."""
+    components: list[Component] = []
+
+    _check_lockfile(request)
+    lockfile_name = request.source_dir.join_within_root(DEFAULT_LOCKFILE_NAME)
+    log.info(f"Reading the rpms lockfile: {lockfile_name}")
+
+    with open(lockfile_name) as f:
+        raw_content = yaml.safe_load(f)
+
+        # Create a list of format handlers and add handler for every supported format
+        # Lockfile processor will use the first handler that matches the input format
+        lockfile_format_processor = LockfileFormat(raw_content)
+        lockfile_format_processor.add_handler(RedhatRpmsLock)
+        lockfile_format_processor.process_formats()
+
+        package_dir = request.output_dir.join_within_root(DEFAULT_PACKAGE_DIR)
+        handler = lockfile_format_processor.matched_handler
+        handler.download(package_dir)
+        handler.verify_downloaded()
+        files_metadata = handler.metadata
+        _generate_sbom_components(components, files_metadata)
+
+    return RequestOutput.from_obj_list(
+        components=components,
+        environment_variables=_generate_environment_variables(DEFAULT_PACKAGE_DIR),
+        project_files=_generate_repofiles(request.output_dir),
+    )
+
+
+def _generate_sbom_components(components: list[Component], files_metadata: dict) -> None:
+    """Fill the compoenent list with the package records."""
+    for file_path, file_metadata in files_metadata.items():
+        # iterate only through package metadata (skip sources metadata)
+        if not file_metadata["package"]:
+            continue
+        # get (optional) 'epoch' - can't get it via rpmfile.headers
+        rpm_args = [
+            "-q",
+            "--queryformat",
+            "%|EPOCH?{%{EPOCH}}:{}|",  # return "" when epoch is not set instead of "(None)"
+            file_path,
+        ]
+        epoch = run_cmd(cmd=["rpm", *rpm_args], params={})
+
+        # NOTE: add license into SBOM?
+        rpm_args = [
+            "-q",
+            "--queryformat",
+            "%{LICENSE}",
+            file_path,
+        ]
+        license = run_cmd(cmd=["rpm", *rpm_args], params={})
+
+        with rpmfile.open(file_path) as rpm:
+            vendor = rpm.headers.get("vendor", b"").decode()
+            vendor = quote(vendor.lower())
+
+            name = rpm.headers.get("name", b"").decode()
+            version = rpm.headers.get("version", b"").decode()
+            release = rpm.headers.get("release", b"").decode()
+            arch = rpm.headers.get("arch", b"").decode()
+            download_url = quote(file_metadata["url"])
+
+            # https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#rpm
+            # https://github.com/package-url/purl-spec/blob/master/PURL-SPECIFICATION.rst#known-qualifiers-keyvalue-pairsa
+            purl = (
+                f"pkg:rpm{'/' if vendor else ''}{vendor}/{name}@{version}{release}"
```

Comment:
'-' is missing between {version} and {release}

Reply:
Here's some unit tests https://github.com/containerbuildsystem/cachi2/compare/main...hlin:cachi2:prefecting-rpms-test